simulation/utils/generate_map.py

```diff
@@ -0,0 +1,172 @@
+import random
+import math
+import matplotlib.pyplot as plt
+from enum import Enum
+import numpy as np
+from scipy.interpolate import BSpline, splprep, splev
+import copy
+
+
+class Difficulty(Enum):
+    """Enum for Difficulty level"""
+    PATHETIC = 0
+    EASY = 1
+    MEDIUM = 2
+    HARD = 3
+    REALLY_HARD = 4
+    WTF = 5
+
+
+def intersect(A, B, C, D):
+    """Helper of is_crossing function
+    Returns True if line segments AB and CD intersect"""
+    ccw = lambda A, B, C: (C[1] - A[1]) * (B[0] - A[0]) > (B[1] - A[1]) * (C[0] - A[0])
+    return ccw(A, C, D) != ccw(B, C, D) and ccw(A, B, C) != ccw(A, B, D)
+
+
+def is_crossing(X, Y, i, j):
+    """Function that checks if two lines ((X[i],Y[i]), (X[i+1], Y[i+1])), ((X[j],Y[j]), (X[j+1],Y[j+1])) intersect"""
+    if i == j:
+        return False
+    try:
+        return intersect((X[i], Y[i]), (X[i + 1], Y[i + 1]), (X[j], Y[j]), (X[j + 1], Y[j + 1]))
+    except IndexError:
+        print(f'i = {i}, j = {j}, size = {len(X)}')
+
+
+def next_point(x_prev, y_prev, step, alpha, y_res):
+    """Function returning next random point for the generalized map"""
+    y_range = 2 * step * math.tan(alpha / 2)
+    ymin = y_prev - y_range / 2
+    ymin = ymin if ymin >= 0 else 0
+    ymax = y_prev + y_range / 2
+    ymax = ymax if ymax < y_res - 20 else y_res - 20
+    p = random.random()
+    if p >= 0.3:
+        x = x_prev + step
+    else:
+        x = x_prev - step if x_prev > step else x_prev + step
+    y = random.uniform(ymin, ymax)
+    return x, y
+
+
+def exterminate_loops(Xold, Yold, step, y_res, ran=40):
+    """Function that gets rid of most of the loops in map so as to make it a little less crazy"""
+    X, Y = copy.deepcopy(Xold), copy.deepcopy(Yold)
+    i = 0
+    size = len(X)
+    while i < size - 1:
+        i = size - 2 if i >= size - 1 else i
+        left_limit = i - ran if i >= ran else 0
+        right_limit = i + ran if i + ran < size else size - 2
+        for j in range(left_limit + 1, right_limit):
+            j = size - 2 if j >= size - 1 else j
+            if is_crossing(X, Y, i, j):
+                del X[i + 1]
+                del Y[i + 1]
+                x, y = next_point(X[size - 2], Y[size - 2], step, math.pi * 3 / 4, y_res)
+                X.append(x)
+                Y.append(y)
+        i += 1
+    return X, Y
+
+
+def prepare_map(x_res, y_res, step, angle_range, starting_y):
+    """Function that prepares real map - interpolates random points generated by function next_point
+        returns interpolated and smoothed map in X and Y arrays
+    """
+    X = []
+    Y = []
+    x = 0
+    y = random.randrange(y_res / 4, y_res * 3 / 5) if not starting_y else starting_y
```

Comment:
Nazwy mało znaczące. No i można to y_res/4 i y_res* 3/5 przypisać do osobnych zmiennych coby było wiadomo co to jest wgl